IPv6 support for Windows and Linux

wltjr opened the pull request:
NOT MY WORK! Updated patches from the following:

From adabfe11d45e9e8fd5bfc445dc02ea80497c308f Mon Sep 17 00:00:00 2001
From: WilliamKyle <williamkyle@126.com>
Date: Thu, 2 Apr 2015 23:45:56 +0800
Subject: [PATCH 1/2] IPv6 support for Windows and Linux

From a4c7ec47ad2447e8bcbb2fa99aac370c0b50d429 Mon Sep 17 00:00:00 2001
From: Fabian Krueger <krueger.fabian@gmail.com>
Date: Sat, 29 Oct 2016 20:21:54 +0200
Subject: [PATCH 2/2] Fix build issue with IpcServer after applying ipv6 patch

PR #4499
https://github.com/symless/synergy-core/pull/4499

Fixes Issue #4325
https://github.com/symless/synergy-core/issues/4325

yupi2 commented on src/lib/net/TCPSocketFactory.h:
```diff
@@ -31,9 +32,9 @@ class TCPSocketFactory : public ISocketFactory {
     // ISocketFactory overrides
     virtual IDataSocket*
-                        create(bool secure) const;
+                        create(bool secure), IArchNetwork::EAddressFamily family = IArchNetwork::kINET) const;
```
need to remove a `)` after `secure` here

yupi2 commented on src/lib/net/TCPSocketFactory.h:
```diff
@@ -31,9 +32,9 @@ class TCPSocketFactory : public ISocketFactory {
+                        create(bool secure), IArchNetwork::EAddressFamily family = IArchNetwork::kINET) const;
     virtual IListenSocket*
-                        createListen(bool secure) const;
+                        createListen(bool secure), IArchNetwork::EAddressFamily family = IArchNetwork::kINET) const;
```
need to remove a `)` after `secure` here

wltjr commented:
@yupi2 thanks! Total fail there... Corrected and force pushed. Please let me know if I missed anything else. I clearly did not try the patches myself :P

yupi2 commented on src/lib/arch/win32/ArchNetworkWinsock.cpp:
```diff
@@ -212,6 +213,11 @@ ArchNetworkWinsock::newSocket(EAddressFamily family, ESocketType type)
     }
     try {
         setBlockingOnSocket(fd, false);
+        OOL flag = 0;
```
`OOL` to `BOOL`

wltjr replied:
Thanks again, corrected and re-force pushed :)

wltjr commented:
@XinyuHou  @nbolton What is going on guys? This is the 2nd PR now for IPv6 being ignored. I updated the files, it was ready to merge. Now there has been another release and the patches must be updated again...

Is IPv6 support something synergy cares to support like ever? Seriously there is no reason to keep avoiding these PRs for years. Almost 3 years now from the date of the initial patches in April 2015. I have been using this with IPv6 for over a year now patched. I really see no reason to keep omitting this, short of it not being wanted. Which seems to be the case.

Otherwise can we get a reason why it is not being merged? I can update the files again. Which will make the 3rd PR or set of patches for IPv6 support in synergy. IPv6 cannot be avoided for ever. Many have been using it for years.

Thanks!

mrsixw commented:
@wltjr Thanks for the PR. We are not deliberately ignoring this by any stretch, but we are a small team and our priorities do vary over time. Part of my remit over the coming month will be to try and establish a more regular cadence for reviewing synergy-core PRs.

wltjr commented:
@yupi2 if you could review as you did before. Pretty sure I did not miss anything but could use another set of eyes. You caught a few before. Thanks!

@mrsixw I rebased and fixed the modifications so its once again ready to merge. Please review and merge ASAP. Thank you!

yupi2 commented:
@wltjr It seems like a lot of recheck / compare so I'm not going to this time.

wltjr commented:
@yupi2 no worries just seemed like you were the only one to review last time. It should be good to go, as I rebased, and only a few files had conflicts. That was mostly where they dropped secure/tls stuff. Much more minor than before. I remade [another that applies against 2.0.0](https://github.com/Obsidian-StudiosInc/synergy-core/commit/e57e0b40953153b6100321f90085c31ff5c22a8c.patch), less windows stuff, so I can patch that version. Which I am running 2.0.0 with that patch right now, no problems. The patches here are against HEAD and do not apply to 2.0.0. The original set works with 1.8.8, but not to many commits beyond that tag.

That's 3 sets of IPv6 patches.... But the one here should hopefully get merged ASAP. I hope I do not have to do this for a 3rd/4th time...

nlyan commented:
@mrsixw IPv6 support in core is not really relevant for us now, since all networking in v2.0 is outside of core. We're only using TCP between core and v2 for legacy reasons.

wltjr commented:
@nlyan people want to use Synergy over IPv6. Where the code goes, be it in core or other is moot. Synergy needs to have support for IPv6, some years ago. Do you all think IPv6 is not a thing? Not in use? IPv4 is here forever?

IPv6 should be relevant somewhere. I care less if its in core or other. Just needs to exist. I am sure others feel the same. PR and bugs have only been open since like 2015...
https://github.com/symless/synergy-core/issues/4325
https://github.com/symless/synergy-core/pull/4499

Really not sure what the hang ups or issue really is. Seems like a stream of excuses or something. Merge the code, move it else where later if your changing the Networking aspect. Though given what synergy is, I am not clear how you are moving networking out of core. Given networking is a core feature of synergy. How else are two computers supposed to exchange information on keyboard and mouse if not via network?

Not to be rude, but it really sounds like you haven't a clue what your talking about. If the network code was moved, the patch would not apply. I still see networking files here in core
https://github.com/symless/synergy-core/tree/master/src/lib/net

Same for 2.0 branch
https://github.com/symless/synergy-core/tree/v2.0.0-stable/src/lib/net

Anyway its moot. I have the patches I need and can continue to modify things for my needs. Clearly synergy developers care less about IPv6. Just be honest. You all either do not want to add IPv6. Lack the knowledge or ability, or something. Thus far just seems like excuses with no merit or factual basis.

If I wasn't updating the patches/PRs and making noise. This would be a non issues going no where even less than it is now. So lets be clear. Synergy does not care about IPv6 support. Thus ignoring the initial PR for years, my 2nd and now 3rd pr, and now just excuse after excuse. For code that can merge per CI, etc.

nlyan commented:
We're not simply ignoring the issue, but there has been very low demand to support IPv6 from customers, and we just haven't reviewed this code for that reason.

Synergy v2 doesn't use the networking code in synergy-core to connect over the LAN, so supporting IPv6 in core has no upside for us. Merging this to *core* at this point would just risk further instability.

Adding IPv6 support to the proprietary components of v2 will be fairly trivial, we just have to take the time to make a few tweaks and turn it on. Again, we've just had much higher priorities since v2.0.0

I might suggest you try and get this merged in to one of the popular open source v1 forks such as https://github.com/debauchee/barrier/ or https://github.com/yupi2/synergy as I have no desire to see your hard work go to waste.

wltjr commented:
@nlyan thanks for the explanation! Not so much my work, it was another's I updated twice. Not seeing much demand is the best explanation. I would assume that maybe because there are less people using multiple systems these days in our new mobile world... Who sits and uses a keyboard and mouse? Old fogies...

If you have proprietary components, you all may want to code the IPv6 support yourself to retain copyright. Unless you have some contributor sign off as to such. Its moot to me, but since the original work is not mine. I cannot really sign off any anything along those lines. Maybe a none issue, but came to mind when you said proprietary. FOSS contributions can get tricky there in such situations.

Thanks for the suggestion on forks! I was not aware of them. I am not so concerned about the work being unappreciated as just moving the ball forward. I like to get stuff done ASAP, so if I can help to expedite etc. That was my interest, greasing the wheels. I do not have much personally invested. Though do not want to endless redo PRs either.

Thanks again for further clarification!

nlyan commented:
Just to clarify, our proprietary codebase is totally separate and doesn't contain any "Synergy core" code.

wltjr commented:
@nlyan will this project continue to have a FOSS version or just move to proprietary? Based on information thus far, seems like aspects of synergy-core will not be furthered and maybe removed, like the network code. Sorry just trying to clarify, I think I am confused now. It sounded like synergy-core was used with the proprietary one. Just different code bases, one building/running on the other. Thus no point in adding IPv6 to that, if it can be added else where.

Is that correct? Is the proprietary one still using synergy-core? Not sharing code, using synergy-core as a layer per se.

I did not realized there was a proprietary version. I never looked into synergy. I just used the package as provided by distro.

nlyan commented:
The proprietary (as of v2.0) product known as Synergy, produced by us at Symless, uses "Synergy Core" (everything in this repository) to support its core feature set. A synergy-core(.exe) binary is shipped with v2.0, and this constitutes an unaltered build from master here.

I can't speak for the business wrt to whether future features will be open sourced or not, but so far only features that have been rewritten/replaced in our new codebase have been removed from master here to signal that we don't intend on supporting them any longer. Any additions to core will be added to the this repo as required to comply with the GPL.

wltjr commented:
@nylan ok that is what I was thinking. Based on Symless needs, if Symless needs IPv6.It would go into the Symless proprietary Synergy products. Thus not adding this code to the underlying synergy-core. That makes sense in that regard. If you all are implementing other features on top of the core.

I understand not being able to say about future features going into synergy-core vs Symless Synergy products. Also makes sense as why some code would be removed from core, and other code like IPv6 support not added to core. I was more asking the other way, which has been answered indirectly. If IPv6 was added, where would it go? IPv6 support would not be added to core but go into Symless Synergy products. I assume further networking code maybe removed from core if not used in Symless Synergy products.

I guess there is not really any road forward for this PR then if that is the case. May need to comment on the bug as to that effect. Which would make this type of PR better suited to a fork as you mentioned.

nlyan commented:
IPv6 support would definitely go in to our proprietary code because that's where the new network layer lives.

wltjr commented:
Yeah so I guess we can close this PR then, as it will be invalid. Unless you all implement it in way that leaves the existing networking stuff. But then you would have code in core you are not using, and not sure about maintaining, etc. Likely should comment on general IPv6 bug that it will be going into Synergy products, not the open source core, as the Network layer was moved. Which I assume the rest of the networking in core that remains will be removed someday.

nlyan commented:
Some kind of networking in core will always be required, whether that be TCP sockets or an IPC, but at the moment that is fine on IPv4 because it only connects on loopback. 

Leaving PR for a bit longer to give forks a chance

wltjr commented:
Ok no worries, at least I understand the situation. I am fine with the PR remaining open for forks, or if till decided otherwise for synergy-core. Thank you very much for your time in clarification!

nbolton commented:
We could merge this in, though it has no advantage to Symless or it's customers (for the reasons that @nlyan highlighted), it may actually have some sort of benefit to the open source community at large (i.e. people that use only synergy-core via the command line that they build themselves).

@mrsixw How much impact (on time and risk) would it be on our dev resources to review this PR and potentially merge it in to benefit non-customers?

mrsixw commented:
@nbolton We'll take some time in our next planning cycle to review this and see where we are. Once reviewed, we'll see about merging.

nbolton commented:
Fixes #4154

srl295 commented:
This PR breaks mDNS (`.local`) at least on mac: server listens on IPv4, but client attempts to connect on IPv6 : #6964
